### PR TITLE
CAMEL-14847: create a camel-jq component

### PR DESCRIPTION
**Pull request by lburgazzoli**

### CODE REVIEW

**Comment by lburgazzoli** [comment]
@davsclaus @oscerd this is a first rough draft of a new language based on JQ (https://issues.apache.org/jira/browse/CAMEL-14847).

As it's been a while since I created a language/component, I'm asking for advice on how to do it properly :)

File: bom/camel-bom/pom.xml [hunk 1]
```diff
@@ -1082,6 +1082,11 @@
         <artifactId>camel-jackson-avro</artifactId>
         <version>${project.version}</version>
       </dependency>
+      <dependency>
+        <groupId>org.apache.camel</groupId>
+        <artifactId>camel-jackson-jq</artifactId>
```
Comment (davsclaus) [inline bom/camel-bom/pom.xml:1087]: Maybe this is wrong as its called `camel-jq` in all the other places

File: components/camel-jq/pom.xml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-support</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>net.thisptr</groupId>
+            <artifactId>jackson-jq</artifactId>
```
Comment (davsclaus) [inline components/camel-jq/pom.xml:54]: Add `version` here

File: components/camel-jq/src/generated/resources/META-INF/services/org/apache/camel/other.properties [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Generated by camel build tools - do NOT edit this file!
```
Comment (davsclaus) [inline components/camel-jq/src/generated/resources/META-INF/services/org/apache/camel/other.properties:1]: I think the other.properties is wrong, maybe try to `rm -rf src/generated` to regen from fresh

File: components/camel-jq/src/main/java/org/apache/camel/language/jq/JqExpression.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(JqExpression.class);
+
+    private final String expression;
+    private final Scope scope;
+
+    private ObjectMapper objectMapper;
+    private String resultTypeName;
+    private Class<?> resultType;
+
+    private boolean autoDiscoverObjectMapper;
```
Comment (davsclaus) [inline components/camel-jq/src/main/java/org/apache/camel/language/jq/JqExpression.java:39]: I wonder if you want/need for these fine grained auto discover functions to be able to turn them on|off individually?

Would you for example turn off fromJq only? Or from serviceLoader only? And are these two built-in functions from JQ itself?

File: core/camel-core-model/src/main/java/org/apache/camel/model/language/JqExpression.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    public void setAutoDiscoverFunctionsFromServiceLoader(String autoDiscoverFunctionsFromServiceLoader) {
+        this.autoDiscoverFunctionsFromServiceLoader = autoDiscoverFunctionsFromServiceLoader;
+    }
+
+    public String isAutoDiscoverFunctionsFromJq() {
+        return autoDiscoverFunctionsFromJq;
+    }
+
+    /**
+     * Set if JQ functions should be discovered from JQ's json file.
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/language/JqExpression.java:99]: What does this mean? Maybe the doc page should have more details? For users that dont know JQ to well. Does it mean you can have built-in functions in a JQ script file?

File: core/camel-core-model/src/main/java/org/apache/camel/model/language/JqExpression.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlTransient;
+
+import org.apache.camel.spi.Metadata;
+
+/**
+ * Evaluates a JQ expression.
+ */
+@Metadata(firstVersion = "3.18.0", label = "language", title = "Jq")
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/language/JqExpression.java:30]: title = `JQ`

File: core/camel-core-model/src/main/java/org/apache/camel/model/language/JqExpression.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlTransient;
+
+import org.apache.camel.spi.Metadata;
+
+/**
+ * Evaluates a JQ expression.
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/language/JqExpression.java:28]: Maybe do like jsonpath:

Evaluates a JQ expression against a JSON message body.

label = "language,json"

**Comment by davsclaus** [comment]
Great work you have not lost the knowledge how to build new languages.

Also maybe add support for `headerName` that some languages have that makes it possible to supply the json in a header instead of message body. See JSonPath language

**Comment by lburgazzoli** [comment]
@davsclaus I've fixed most of the findings (doc is still not up to date)

One additional question I have is that in the camel-jackson component, the Jackson Type converter [must explicitly be enabled ](https://github.com/apache/camel/blob/main/components/camel-jackson/src/main/java/org/apache/camel/component/jackson/converter/JacksonTypeConverters.java#L72).

I wonder if some basic conversion (i.e. InputStream/String/byte[]/Map <--> JsonNode) can be enabled by default.
The reason for that is that the jq extension requires the input to be a JsonNode and of course I can do my own set if/else but maybe it would be better to leverage the type converter, WDYT ? as alternative the doc can mention that the JacksonTypeConverters shuld be explicitly enabled.

**Comment by davsclaus** [comment]
Ah yeah the problem is that its a fallback type converter that is too generic.

We can add explict type converters that uses `JSonObject` `JsonNode` or what there is that are 100% json then we can use regular type converters that are always enabled.

And we can modify the fallback to be for special use-cases.

You are welcome to create a JIRA about this

**Comment by lburgazzoli** [comment]
> 
> You are welcome to create a JIRA about this

https://issues.apache.org/jira/browse/CAMEL-18165

**Comment by davsclaus** [comment]
Out of the box converters added

**Review by davsclaus — APPROVED** [review]

**Review by oscerd — APPROVED** [review]

**Review by davsclaus — APPROVED** [review]

**Pull request merged**